### PR TITLE
[Python/SQLAlchemy] Add example programs about efficient bulk inserts with SQLAlchemy, pandas, and Dask

### PR DESCRIPTION
**Pull request by amotl**
Hi there,

this patch carries over the example programs about Python/SQLAlchemy/pandas/Dask from https://github.com/crate/crate-python/pull/552. Originally, the corresponding programs have been conceived as individual gists:

- [SQLAlchemy insert gist]
- [pandas insert gist]
- [Dask insert gist]

With kind regards,
Andreas.

/cc @marijaselakovic, @karynzv, @WalBeh, @hammerhead, @hlcianfagna, @andnig, @ckurze 

[Dask insert gist]: https://gist.github.com/amotl/c8039291fb4b4b16879fd8814de9ecdf
[pandas insert gist]: https://gist.github.com/amotl/d39db8eff6ad5de5898ce61fac4c0af3
[SQLAlchemy insert gist]: https://gist.github.com/amotl/d0fa29ac52d4c10e4a2544cffd606dfc

### Backlog
- [x] ~~Software tests and Dependabot configuration works well for the other code example directories, and will be added here as well.~~ _Done, with a code coverage of 95%._
- [x] ~~https://github.com/crate/crate-python/pull/553 needs to be merged and released beforehand, to make the examples actually work.~~ _Merged in July 2023._
- [ ] Some adjustments may be added to bring all programs into the same shape.
- [ ] Other than demonstrating write operations, also demonstrate read operations?

=> Backlog items have been added to GH-95, in order to finally get this merged, so it can be promoted and re-used.

### CODE REVIEW

File: .github/workflows/test-python-sqlalchemy.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+          python-version: ${{ matrix.python-version }}
+          architecture: x64
+          cache: 'pip'
+          cache-dependency-path: |
+            by-language/python-sqlalchemy/requirements.txt
+            by-language/python-sqlalchemy/requirements-dev.txt
+
+      - name: Validate by-language/python-sqlalchemy
+        run: |
+          python testing/ngr.py --accept-no-venv by-language/python-sqlalchemy
```
Comment (amotl) [inline .github/workflows/test-python-sqlalchemy.yml:62]: We finally added a universal test runner, effectively just wrapping a few other calls, to be able to maintain a concise incantation syntax.

File: testing/ngr.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+-----
+About
+-----
+
+Next Generation Runner (ngr): Effortless invoke programs and test harnesses.
+
+-------
+Backlog
+-------
+- After a few iterations, refactor to separate package.
```
Comment (amotl) [inline testing/ngr.py:11]: `ngr.py` will only be incubated here, it will be refactored to a standalone package after a few iterations.

File: by-language/python-sqlalchemy/README.rst [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+############################################################
+Connect to CrateDB and CrateDB Cloud using Python/SQLAlchemy
+############################################################
+
+
+*****
+About
+*****
+
+Example programs demonstrating CrateDB's SQLAlchemy adapter and dialect.
```
Comment (amotl) [inline by-language/python-sqlalchemy/README.rst:12]: Currently, it is mostly about pandas/Dask operations through SQLAlchemy. This is a special topic, where we also added extensive documentation about, at [^1]. Here, we surely want to link to it.

[^1]: https://cratedb.com/docs/python/en/latest/by-example/sqlalchemy/dataframe.html

File: by-language/python-sqlalchemy/README.rst [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+
+*****
+Usage
+*****
+
```
Comment (amotl) [inline by-language/python-sqlalchemy/README.rst:18]: ### Improve documentation II

In this case, within the **Usage** section, it should clearly be focused on "library use" details, like, how to exactly use the `insert_bulk` function with pandas and Dask.

```python
from crate.client.sqlalchemy.support import insert_bulk

# pandas
df.to_sql(name=self.table_name, con=engine, if_exists="append", index=False, chunksize=bulk_size, method=insert_bulk)

# Dask
ddf.to_sql("testdrive", uri=DBURI, index=False, if_exists="replace", chunksize=10_000, parallel=True, method=insert_bulk)
```

What is currently within the **Usage** section, should go into an **Examples** section instead.

File: by-language/python-sqlalchemy/README.rst [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    time python insert_efficient.py cratedb batched
+
+    time python insert_pandas.py
+    time python insert_pandas.py --mode=basic --insertmanyvalues-page-size=5000
+    time python insert_pandas.py --mode=bulk --bulk-size=20000 --num-records=75000
+
+    time python insert_dask.py
+
+    # Connect to CrateDB Cloud.
+    time python insert_pandas.py --dburi='crate://admin:<PASSWORD>@example.aks1.westeurope.azure.cratedb.net:4200?ssl=true'
```
Comment (amotl) [inline by-language/python-sqlalchemy/README.rst:46]: This is a bit poor. Improve/highlight operations with CrateDB Cloud.

File: by-language/python-sqlalchemy/README.rst [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    docker run -it --rm --publish=4200:4200 --publish=5432:5432 crate
+
+Navigate to example program directory, and install prerequisites::
+
+    # Acquire sources.
+    git clone https://github.com/crate/cratedb-examples
+    cd cratedb-examples/by-language/python-sqlalchemy
+    python3 -m venv .venv
+    source .venv/bin/activate
+    pip install --upgrade --requirement requirements.txt
```
Comment (amotl) [inline by-language/python-sqlalchemy/README.rst:30]: A test suite has been added. Document how it can be invoked.

```
pytest
```

File: by-language/python-sqlalchemy/insert_pandas.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+import pkg_resources
+import sqlalchemy as sa
+from colorlog.escape_codes import escape_codes
+from pandas._testing import makeTimeDataFrame
+
+from crate.client.sqlalchemy.support import insert_bulk
+
+logger = logging.getLogger(__name__)
+
+pkg_resources.require("sqlalchemy>=2.0")
```
Comment (amotl) [inline by-language/python-sqlalchemy/insert_pandas.py:60]: Also reflect this within the `requirements.txt` file, and probably also within `insert_dask.py`.

File: testing/ngr.py (hunk 2 above)
Comment (amotl) [reply testing/ngr.py:11]: `testing/ngr.py` moved into `pueblo.ngr` already, and learned three more directory shapes to invoke: `javascript`, `make`, and `rust`.

```python
class ItemType(Enum):
    DOTNET = "dotnet"
    JAVA = "java"
    JAVASCRIPT = "javascript"
    MAKE = "make"
    PHP = "php"
    PYTHON = "python"
    RUBY = "ruby"
    RUST = "rust"
```

- https://github.com/pyveci/pueblo/pull/6/commits/b7a25308ace20

File: by-language/python-sqlalchemy/README.rst (hunk 3 above)
Comment (amotl) [reply by-language/python-sqlalchemy/README.rst:12]: The README has been improved, see its rendered version here.

-- https://github.com/crate/cratedb-examples/blob/amo/python-sqlalchemy/by-language/python-sqlalchemy/README.rst

File: by-language/python-sqlalchemy/README.rst (hunk 4 above)
Comment (amotl) [reply by-language/python-sqlalchemy/README.rst:23]: Fixed, see above.

File: by-language/python-sqlalchemy/README.rst (hunk 5 above)
Comment (amotl) [reply by-language/python-sqlalchemy/README.rst]: Dito. Fixed.

File: by-language/python-sqlalchemy/README.rst (hunk 6 above)
Comment (amotl) [reply by-language/python-sqlalchemy/README.rst:75]: Fixed.

File: by-language/python-sqlalchemy/insert_pandas.py (hunk 7 above)
Comment (amotl) [reply by-language/python-sqlalchemy/insert_pandas.py:60]: Fixed.

File: by-language/python-sqlalchemy/insert_efficient.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    # Run PostgreSQL
+    docker run --rm -it --publish=5432:5432 --env "POSTGRES_HOST_AUTH_METHOD=trust" postgres:15 postgres -c log_statement=all
+
+    # Use SQLite
+    time python insert_efficient.py sqlite multirow
+    time python insert_efficient.py sqlite batched
+
+    # Use PostgreSQL
+    time python insert_efficient.py postgresql multirow
+    time python insert_efficient.py postgresql batched
```
Comment (seut) [inline by-language/python-sqlalchemy/insert_efficient.py:36]: How is this example related to other DB's? Why should I run this for other DB's?

File: by-language/python-sqlalchemy/README.rst [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+Example programs demonstrating CrateDB's SQLAlchemy adapter and dialect.
+
+
+*****
+Usage
+*****
+
+To start a CrateDB instance on your machine for evaluation purposes, invoke::
+
+    docker run -it --rm --publish=4200:4200 --publish=5432:5432 crate
```
Comment (seut) [inline by-language/python-sqlalchemy/README.rst:66]: Is publishing the PG port `5432` required for this example?

File: by-language/python-sqlalchemy/insert_pandas.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    logging.basicConfig(format=log_format, level=level, handlers=[handler])
+
+    # Enable SQLAlchemy logging.
+    if SQLALCHEMY_LOGGING:
+        logging.getLogger("sqlalchemy").setLevel(level)
+
+
+@click.command()
+@click.option("--dburi", type=str, default="crate://localhost:4200", required=False, help="SQLAlchemy database connection URI.")
+@click.option("--mode", type=str, default="bulk", required=False, help="Insert mode. Choose one of basic, multi, bulk.")
```
Comment (seut) [inline by-language/python-sqlalchemy/insert_pandas.py:142]: could use a choice option to make possible options strict.
```suggestion
@click.option("--mode", type=click.Choice(['basic', 'multi', 'bulk']), default="bulk", required=False, case_sensitive=False, help="Insert mode. Choose one of basic, multi, bulk.")
```

**Review by seut — APPROVED** [review]

File: by-language/python-sqlalchemy/insert_pandas.py (hunk 7 above)
Comment (seut) [reply by-language/python-sqlalchemy/insert_pandas.py:60]: Is it still needed at all then?

File: by-language/python-sqlalchemy/insert_efficient.py (hunk 8 above)
Comment (amotl) [reply by-language/python-sqlalchemy/insert_efficient.py:36]: I wanted to have an easy stage to run comparisons side-by-side. It is now integrated into the relevant example program, and I find it convenient.

File: by-language/python-sqlalchemy/README.rst (hunk 9 above)
Comment (amotl) [reply by-language/python-sqlalchemy/README.rst:66]: Erm. Probably not. It is just my canonical command I slap everywhere, not needing to discriminate between different variants, and also advertising a bit

File: by-language/python-sqlalchemy/insert_pandas.py (hunk 10 above)
Comment (amotl) [reply by-language/python-sqlalchemy/insert_pandas.py:142]: Did you ask another expert on this?
```
TypeError: Parameter.__init__() got an unexpected keyword argument 'case_sensitive'
```
